### ISSUE TITLE
Building from tarballs without Wayland support fails

### ISSUE THREAD

**Issue by alanc** [ibus/ibus#2609]
Please fill in the following items if you don't know the root cause.

Which distribution and version?: 
Solaris 11.4

Which desktop environment and version?: 
GNOME 41

Which session type?: 
X11

IBus version?: 
1.5.29-rc2

Issue description:
Building from the 1.5.29-rc2 tarballs failed with --disable-wayland fails because the distributed *.c files in ui/gtk3 were built with Wayland in use.

Errors include:
```
ibus-1.5.29-rc2/ui/gtk3/application.c:33:10: fatal error: ibuswaylandim.h: No such file or directory
   33 | #include <ibuswaylandim.h>
      |          ^~~~~~~~~~~~~~~~~

ibus-1.5.29-rc2/ui/gtk3/candidatepanel.c:36:10: fatal error: gdk/gdkwayland.h: No such file or directory
   36 | #include <gdk/gdkwayland.h>
      |          ^~~~~~~~~~~~~~~~~~

ibus-1.5.29-rc2/ui/gtk3/panel.c:38:10: fatal error: gdk/gdkwayland.h: No such file or directory
   38 | #include <gdk/gdkwayland.h>
      |          ^~~~~~~~~~~~~~~~~~

ibus-1.5.29-rc2/ui/gtk3/switcher.c:40:10: fatal error: gdk/gdkwayland.h: No such file or directory
   40 | #include <gdk/gdkwayland.h>
      |          ^~~~~~~~~~~~~~~~~~

```

To make it work I had to add to my build recipe:
```
touch ibus-1.5.29-rc2/ui/gtk3/*.vala
```
so that gmake recreated the *.c files from the *.vala sources.

**Comment by fujiwarat** [comment]
Please run `make -C ui/gtk3 maintainer-clean-generic` before you run `make`.
Otherwise I have no idea. The default *.c files are generated with Wayland.

**Comment by alanc** [comment]
That works as well - I'd tried `maintainer-clean` which also removes the Makefile, but didn't know about `maintainer-clean-generic`, thanks.

**Comment by eli-schwartz** [comment]
> Please run `make -C ui/gtk3 maintainer-clean-generic` before you run `make`.
> Otherwise I have no idea. The default *.c files are generated with Wayland.

It is effectively impossible to build ibus correctly, then.

The vala preprocessor (any use of `#if XXXX` in a .vala file evaluates all `--define`s as part of emitting *.c code, which means it is simply wrong to distribute those files in `make distcheck`. But automake insists on adding them.

I tried to hack around this with a dist hook. But I cannot even get `make dist` to run at all:

```
make[5]: Entering directory '/home/eschwartz/git/gentoo/app-i18n/ibus/ibus/bindings/vala'
make  distdir-am
make[6]: Entering directory '/home/eschwartz/git/gentoo/app-i18n/ibus/ibus/bindings/vala'
make[6]: *** No rule to make target 'candidatearea.vala', needed by 'libibus_emoji_dialog_1_0_la_vala.stamp'.  Stop.
make[6]: Leaving directory '/home/eschwartz/git/gentoo/app-i18n/ibus/ibus/bindings/vala'
make[5]: *** [Makefile:906: distdir] Error 2
```


candidatearea.vala is referenced in bindings/vala/Makefile.am but doesn't exist in git. It does exist in the release tarball though. :/

**Comment by eli-schwartz** [comment]
Filed an upstream issue as https://gitlab.gnome.org/GNOME/vala/-/issues/1580

**Comment by fujiwarat** [comment]
> make[6]: *** No rule to make target 'candidatearea.vala', needed by 'libibus_emoji_dialog_1_0_la_vala.stamp'.  Stop.

I cannot reproduce your issue.
`ibus/bindings/vala/Makefile.am` includes `$(LN_S) $(top_srcdir)/ui/gtk3/$@ .`  to generate candidatearea.vala.

`--disable-wayland` works fine:
https://github.com/ibus/ibus/actions/runs/13065223373/job/36456314129

**Comment by eli-schwartz** [comment]
Thanks for pointing that out. Ufortunately, it doesn't work the way you think it does.


More specifically, Makefile.in has this, which helps illuminate the issue:

```
# per file setting is needed to avoid conflicting LN_S by calling
# duplicated times in parallel make
@ENABLE_EMOJI_DICT_TRUE@%.vala: $(ibus_vapi)
@ENABLE_EMOJI_DICT_TRUE@        if test ! -f $@ ; then                                              \
@ENABLE_EMOJI_DICT_TRUE@            $(LN_S) $(top_srcdir)/ui/gtk3/$@ .;                             \
@ENABLE_EMOJI_DICT_TRUE@        fi;
```

The section you are pointing to does exist, but it is ifdefed out in Makefile.am if you build without emoji-dict support, which I did. That means the configure script comments out the whole thing.

**Comment by fujiwarat** [comment]
> The section you are pointing to does exist, but it is ifdefed out in Makefile.am 

Then it would be good to open another issue since your issue is relative with `--disable-emoji-dict`. Seems you don't mention all the conditions.
